Multiprocessing support for NT

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110404
* __->__ #110292

Fixes #110161

Allows NTs to be used in DataLoaders with `num_workers > 1`.

jbschlosser committed:
Multiprocessing support for NT

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

cpuhrsch commented:
Failures look real unfortunately @jbschlosser

jbschlosser committed:
Update on "Multiprocessing support for NT"


Fixes #110161

Allows NTs to be used in DataLoaders with `num_workers > 1`.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

albanD commented on torch/multiprocessing/reductions.py:
```diff
@@ -118,6 +118,20 @@ def rebuild_tensor(cls, storage, metadata):
     return t
 
 
+def rebuild_tensor_nested(cls, storage, metadata):
```
Can we re-use the one from serialization.py?

jbschlosser replied:
For CUDA NTs, I think we need to use the CUDA sharing logic for all this to work properly. I updated the PR with proper CUDA tests and changed the logic here to recursively call into `reduce_tensor()` for the NT components -> rebuild appropriately. Please lmk if you think there is some reuse opportunity I'm missing after this.

jbschlosser commented:
@pytorchbot revert

jbschlosser commented:
@pytorchbot revert -m "Address review comments"

jbschlosser commented:
@pytorchbot revert -m "Address review comments" -c "weird"

jbschlosser committed:
Update on "Multiprocessing support for NT"


Fixes #110161

Allows NTs to be used in DataLoaders with `num_workers > 1`.

[ghstack-poisoned]

albanD commented on torch/multiprocessing/reductions.py:
```diff
@@ -336,6 +343,48 @@ def reduce_tensor(tensor):
+
+
+def reduce_nested_tensor(nt):
+    rebuild_buffer_func, rebuild_buffer_args = reduce_tensor(nt.values().clone())
```
Why clone() ?

jbschlosser replied:
Without the clone():
```
RuntimeError: Attempted to send CUDA tensor received from another process; this is not currently supported. Consider cloning before sending.
```

albanD replied:
When you do what?

jbschlosser replied:
When serializing, `reduce_tensor()` is called on the NT buffer, and `_share_cuda()` for the storage breaks:
```
Traceback (most recent call last):
  File "/home/jbschlosser/anaconda3/lib/python3.9/multiprocessing/queues.py", line 245, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/home/jbschlosser/anaconda3/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/home/jbschlosser/branches/nt_ops/torch/multiprocessing/reductions.py", line 298, in reduce_tensor
    return reduce_nested_tensor(tensor)
  File "/home/jbschlosser/branches/nt_ops/torch/multiprocessing/reductions.py", line 364, in reduce_nested_tensor
    rebuild_buffer_func, rebuild_buffer_args = reduce_tensor(nt.values())
  File "/home/jbschlosser/branches/nt_ops/torch/multiprocessing/reductions.py", line 310, in reduce_tensor
    ) = storage._share_cuda_()
  File "/home/jbschlosser/branches/nt_ops/torch/storage.py", line 1073, in _share_cuda_
    return self._untyped_storage._share_cuda_(*args, **kwargs)
RuntimeError: Attempted to send CUDA tensor received from another process; this is not currently supported. Consider cloning before sending.
```

albanD replied:
But is that because you try to send stuff that you received from another process as the error suggests? :p

jbschlosser replied:
Haha probably, but I'm unclear as to why that's happening. This is from the test case in this PR that operates with `batch_size=1` and `num_workers>0`. I'd expect each NT in the dataset to be sent only once to one of the workers, but something more complicated than this is happening.

jbschlosser replied:
Okay I figured it out: the worker threads are sending stuff they received from the parent thread due to the use of `_identity()`. I'm going to remove the `clone()` and fix the test case.

albanD commented on torch/multiprocessing/reductions.py:
```diff
@@ -290,6 +290,13 @@ def reduce_tensor(tensor):
+    # TODO: Handle distinguishing between subclass and non-subclass versions of NT better
+    from torch.nested._internal.nested_tensor import NestedTensor
+
+    if tensor.is_nested and not isinstance(tensor, NestedTensor):
```
The python version also returns True for is_nested?

jbschlosser commented on test/test_dataloader.py:
```diff
@@ -2266,6 +2272,22 @@ def test_excessive_thread_creation_warning(self):
+        )
+
+        for i, batch in enumerate(loader):
+            self.assertEqual(batch[0], dataset[i])
```
The input to the `collate_fn` is a batch-sized list of things to collate, so identity returns this list.

jbschlosser commented on test/test_dataloader.py:
```diff
@@ -2266,6 +2272,22 @@ def test_excessive_thread_creation_warning(self):
+            batch_size=1,
+            num_workers=4,
+            collate_fn=_identity,
+            multiprocessing_context=('spawn' if 'cuda' in device else 'fork'),
```
for cuda, I believe only spawn and forkserver work

jbschlosser commented on test/test_dataloader.py:
```diff
@@ -2266,6 +2272,22 @@ def test_excessive_thread_creation_warning(self):
+            dataset,
+            batch_size=1,
+            num_workers=4,
+            collate_fn=_identity,
```
Sadly no, there is some dense-specific logic that does concatenation directly into shared memory. NT doesn't support `nt.new(storage)` at the moment:
```
RuntimeError: Unrecognized tensor type ID: NestedTensorCUDA
```

Alternatively, we could add nested support to the default `collate_fn`?

jbschlosser replied:
Conceptually, there's two ways I can see the default `collate_fn` being expanded:
1. **Input batch list contains dense tensors of ragged sizes**: produce a NT output instead of `torch.stack()`ing. This is a potentially disruptive change so it may be better to require an explicit `collate_fn` override to do this at first. We could throw a descriptive error that points out the raggedness and suggests passing a `collate_fn` that returns a NT.
2. **Input batch list contains NTs** : should the default behavior be to concatenate the list of NTs into a dim+1 NT? This is the case described in the linked issue, although they're implementing a `collate_fn` that just returns the list directly.

jbschlosser committed:
Update on "Multiprocessing support for NT"


Fixes #110161

Allows NTs to be used in DataLoaders with `num_workers > 1`.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Multiprocessing support for NT"


Fixes #110161

Allows NTs to be used in DataLoaders with `num_workers > 1`.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "Causes CUDA memory leaks" -c nosignal

> RuntimeError: CUDA driver API confirmed a leak in __main__.TestDataLoaderDeviceTypeCUDA.test_nested_tensor_multiprocessing_context_forkserver_cuda! Caching allocator allocated memory was 5120 and is now reported as 10240 on device 0. CUDA driver allocated memory was 340459520 and is now 342556672.

https://github.com/pytorch/pytorch/actions/runs/6425541384/job/17449001020

jbschlosser committed:
Update on "Multiprocessing support for NT"


Fixes #110161

Allows NTs to be used in DataLoaders with `num_workers > 1`.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
